Feature/22 implement abstractencoderdecoder and huffmanencoderdecoder

dan1994 opened the pull request:
Closes #22 
@Roynecro97, this is a big one, so I won't merge it immediately, and give you some time to go over it (you may comment and tell me if you intend to, or when you are done).
Also, you may note that design changes are listed in the issue description (#22).
Known bug described in #36.

dan1994 committed:
Added header files

dan1994 committed:
Fixed design to match the current workaround of having changed cannonicalMapping from constexpr to const (see #23)

dan1994 committed:
Made compiler output colored messages

dan1994 committed:
Implemented HuffmanNode methods, and the getFrequencies methods. Changed some declarations and updated design to match

dan1994 committed:
Decided to use a vector of unique ptrs with heap allocations for the frequencies becuase of lifetime issues. To keep code readable and simple, refactored HuffmanNode to Node and added NodePtr and NodeVector aliasses

dan1994 committed:
Refactored getFrequencies test to be more correct

dan1994 committed:
Implemented the buildPrefixlessTree function

dan1994 committed:
Added testing of the buildPrefixlessTree method

dan1994 committed:
Implemented createMapping. Added another static private dfs function which is a recursive helper

dan1994 committed:
Replaced generate to a simple for loop because the vector's size is not incremented unless we call push_back, and the loop is really simple

dan1994 committed:
Fixed upper_bound conditional (it was reversed)

dan1994 committed:
When copying replaced simple iterator with back_inserter, for the list to be aware of the copied elements

dan1994 committed:
WIP in testing buildPrefixlessTree

dan1994 committed:
Included hash implementation for std::pair

dan1994 committed:
Fixed bug in finding the position to insert newly created nodes

dan1994 committed:
Now passing test for all stages of producing a huffman code from string!

dan1994 committed:
Redesigned AbstractEncoderDecoder to be a template class, such that HuffmanEncoderDecoder inherits its specialization. Also made it an interface class (no members)

dan1994 committed:
Changed spec to have padding at the end of the encoded text, for ease of implementation. Added implementation to all member functions except decode (dummy implementation). Changed the interface of encode and decode to be able to report failure

dan1994 committed:
Added decode implementation

dan1994 committed:
Added encode/decode tests and fixed implementation bugs

dan1994 committed:
Added testing for makeEncodingFromText and for encode/decode for random string -> produces segfaults occasionally...

dan1994 committed:
Fixed sefgault by reversing the order of 2 unrelated code blocks. It seems that for some reason, after a try catch block catches an exception, it skips the next chunk of code. I don't understand this behavior

dan1994 committed:
Reformated using clang-format

dan1994 committed:
Added the possibility of failing in creating an encoding from text. Didn't test because the simplest case would involve an input string of size 8.5GB. This is because the easiest way to create a tree of depth 32 is to give each character a frequency of a power of 2. That way, each time we build the next node as a parent of the last built node and a character node, practically increasing the depth of the tree. But that means that for depth of 32, we need sum(2^i) from i=0 to i=31 characters in the string which sums up to 8.5GB

dan1994 committed:
Adapted design code to the conceptual merge between useCanonicalEncoding and makeEncodingFromText (see issue discussion for more details)

dan1994 committed:
Regenerated png from puml

Roynecro97 commented on include/EncoderDecoder/AbstractEncoderDecoder.h:
```diff
@@ -0,0 +1,57 @@
+	 * encoding succeeded and the result of the encoding operation. If the first
+	 * value is false, the value of the second is undefined.
+	 */
+	virtual std::pair<bool, EncodedType> encode(
```
Why return a pair of `bool` and value instead of returning a value and throwing on error?

dan1994 replied:
I wasn't sure about that. I also think in retrospect that throwing may be more appropriate, since there are several failure cases.
This is a non-trivial change (that is relevant in several places), so I opened an issue for it (#39).

Roynecro97 commented on include/EncoderDecoder/AbstractEncoderDecoder.h:
```diff
@@ -0,0 +1,57 @@
+	 * @return true If the encoding was successfuly created
+	 * @return false If failed in creating the encoding
+	 */
+	virtual bool makeEncodingFromText(const std::string &plaintext) = 0;
```
Should the `AbstractEncoderDecoder` be aware of the existence of internal encodings and mapping? At first glance it looks like implementation detail and not part of the interface

dan1994 replied:
If you read the requirements and the intended usage, then there's a clear distinction between using a canonical encoding (which is constructed from the text), and using an external encoding. This API allows this distinction.
There is the option of constructing the encoding from the text elsewhere and then passing it through `setEncoding`. Do you think it will be wiser? If so where should this code be in?

Roynecro97 replied:
I only meant It seems weird that the “abstract” interface knows about the Huffman encoding mechanism, the existence of this API in the concrete class is fine

dan1994 replied:
I think it is an integral part of the API, that applies to any encoding type, so I leave it as is.

Roynecro97 commented on include/EncoderDecoder/Huffman/HuffmanEncoderDecoder.h:
```diff
@@ -0,0 +1,101 @@
+	using EncodedType = PaddedString;
+	using NodePtr = std::unique_ptr<Node>;
+	using NodeVector = std::vector<NodePtr>;
+	friend class HuffmanEncoderDecoder_MakeEncodingFromTextSubRoutines_Test;
```
Maybe move this to the bottom? The idea is to put the class’s interface first and internal stuff later, because the user of the class doesn’t usually care about the internals

dan1994 replied:
I use these aliases in the function declaration for convenience. Is that a good enough excuse?

Roynecro97 replied:
I meant only the friend test class

dan1994 replied:
TODO

Roynecro97 commented on include/EncoderDecoder/Huffman/HuffmanEncoderDecoder.h:
```diff
@@ -0,0 +1,101 @@
+	friend class HuffmanEncoderDecoder_MakeEncodingFromTextSubRoutines_Test;
+
+private:
+	static constexpr int MAX_ENCODED_BITS = 32;
```
Why 32?

dan1994 replied:
This is part of the requirements.
The `.dane` file stores encodings in a way that 32 bits is the max encoding size.
Currently this affects the `MappingType` which maps a `char` to a `std::pair<unsigned char, unsigned int>` (`pair.first` is the size, and `pair.second` is the encoded character).
If I were to disregard this, I might get an integer overflow when creating a mapping.

Roynecro97 commented on include/EncoderDecoder/Huffman/HuffmanEncoderDecoder.h:
```diff
@@ -0,0 +1,101 @@
+
+public:
+	// Inherited methods
+	std::pair<bool, EncodedType> encode(const std::string &plaintext) const;
```
I prefer to state the `virtual` (because it is virtual even if you don’t write it) and always using `override`

dan1994 replied:
TODO

Roynecro97 commented on include/EncoderDecoder/Huffman/PaddedString.h:
```diff
@@ -0,0 +1,12 @@
+
+#include <string>
+
+class PaddedString {
```
`struct` here as a convention for simple data structures

dan1994 replied:
TODO

Roynecro97 commented on include/EncoderDecoder/Huffman/PaddedString.h:
```diff
@@ -0,0 +1,12 @@
+
+class PaddedString {
+public:
+	uint8_t paddingSize;
```
Put this data member after the other for less padding inside your struct

dan1994 replied:
TODO

Roynecro97 commented on src/EncoderDecoder/Huffman/HuffmanEncoderDecoder.cpp:
```diff
@@ -0,0 +1,252 @@
+
+	// Initialize frequencies
+	NodeVector frequencies;
+	frequencies.reserve(256);
```
Magical 256

Roynecro97 replied:
Also, I think this can be done in one call using vector’s ctor
```C++
vector vec(100);
```

dan1994 replied:
> Also, I think this can be done in one call using vector’s ctor
> 
> ```c++
> vector vec(100);
> ```

In this case, I didn't want ctors to be called, but only for space to be reserved.

dan1994 replied:
> Magical 256

TODO

Roynecro97 commented on src/EncoderDecoder/Huffman/HuffmanEncoderDecoder.cpp:
```diff
@@ -0,0 +1,252 @@
+	// Count frequency of each char in plaintext
+	std::for_each(plaintext.cbegin(), plaintext.cend(), [&](unsigned char c) {
+		auto &fp = frequencies[c];
+		(*fp)++;
```
It doesn’t matter here, but I think it’s better to prefer the prefix ++ over the postfix one

dan1994 replied:
This requires adding another function to the `Node` interface which isn't important enough right now, so I'll pass.

Roynecro97 commented on src/EncoderDecoder/Huffman/HuffmanEncoderDecoder.cpp:
```diff
@@ -0,0 +1,252 @@
+
+	// Copy to list non-zero frequency elements
+	std::list<NodePtr> frequenciesList;
+	std::copy(std::make_move_iterator(activeSymbols),
```
Might be because I’m not familiar enough with all algorithms yet but why `copy(move_iterator` instead of the `move()` algorithm?

dan1994 replied:
I didn't know `std::move` had an overload similar to other STL algorithms, but from what I've seen [here](https://stackoverflow.com/a/26433162/11932596), it does.

TODO

Roynecro97 commented on src/EncoderDecoder/Huffman/HuffmanEncoderDecoder.cpp:
```diff
@@ -0,0 +1,252 @@
+		return nullptr;
+	}
+
+	// While there is more than one element in the list
```
Express in comments things that cannot be expressed in code, the `while` itself says the loop’s condition, but what does this loop achieve?

dan1994 replied:
TODO

Roynecro97 commented on src/EncoderDecoder/Huffman/HuffmanEncoderDecoder.cpp:
```diff
@@ -0,0 +1,252 @@
+}
+
+std::pair<bool, HuffmanEncoderDecoder::MappingType>
+	HuffmanEncoderDecoder::createMapping(const NodePtr &root) {
```
The typedefs confuse me, is this a `const unique_ptr&`?

dan1994 replied:
Yes, why?
A note: This function generates a `MappingType` object from the given tree, and discards the tree, so this signature makes sense (at least to me).

Roynecro97 replied:
Passing a reference means there’s no ownership transfer, which makes the `unique_ptr` redundant and forces the user to use one, if you do not intend to transfer ownership to your function, use a raw pointer. If you do, you need the pointer by value, this way it will be moved inside the function

dan1994 replied:
You're right! Might be the alias confused me too! It should definitely pass by value.

TODO

Roynecro97 commented on src/EncoderDecoder/Huffman/HuffmanEncoderDecoder.cpp:
```diff
@@ -0,0 +1,252 @@
+
+	// If the tree is empty, then so should the mapping be
+	if(root == nullptr) {
+		return {false, mapping};
```
But is it a failure?

dan1994 replied:
I guess not. It's probably not what the user intended to happen...
What is your opinion?

Roynecro97 replied:
I think allowing the user to check for an empty mapping and treating this as ok is the right way to go, since this will happen if an empty file is given to your program, and it’s not stated as “not ok” in the readme (as far as I remember)

dan1994 replied:
Seems legit.

TODO

Roynecro97 commented on src/EncoderDecoder/Huffman/HuffmanEncoderDecoder.cpp:
```diff
@@ -0,0 +1,252 @@
+			root,
+			std::make_pair<unsigned char, unsigned int>(0, 0),
+			32);
+	} catch(std::length_error) { return {false, mapping}; }
```
I remember something about using references in `catch`. Otherwise the catch will copy the exception, which means copy the string it holds, which may throw an exception

dan1994 replied:
TODO

Roynecro97 commented on src/EncoderDecoder/Huffman/HuffmanEncoderDecoder.cpp:
```diff
@@ -0,0 +1,252 @@
+std::pair<bool, HuffmanEncoderDecoder::EncodedType>
+	HuffmanEncoderDecoder::encode(const std::string &plaintext) const {
+
+	constexpr unsigned char BITS_IN_CHAR = 8;
```
`std::numeric_limits<char>::digits`

dan1994 replied:
Cool, this is also relevant in a couple of other places.

TODO

dan1994 committed:
Changed PaddedString to struct and reversed internal order to decrease padding size

dan1994 committed:
Added virtual specifier for methods in derived class

dan1994 committed:
Replaced magic numbers

dan1994 committed:
Changed std::copy invocation to std::move in buildPrefixlessTree

dan1994 committed:
Refactored comment to increase readability

dan1994 committed:
All exceptions are now caught by reference

dan1994 committed:
Added override specifier to virtual functions

dan1994 commented:
@Roynecro97, I fixed what I could from your comments, and resolved conversations related to what I fixed.
Please review my comments to unresolved conversations.

Roynecro97 commented:
@dan1994 I replied in the unresolved conversation. I hope the replies are there, the GitHub mobile API is not friendly... :/

dan1994 commented:
Got your replies! Thanks @Roynecro97 !

dan1994 committed:
Moved friend test class to end HuffmanEncoderDecoder declaration

dan1994 committed:
Changed signature of createMapping to receive a non-ref const NodePtr

dan1994 committed:
createMapping doesn't fail on an empty mapping

dan1994 committed:
Updated UML png

dan1994 committed:
Fixed tests to match new createMapping API
